runtime/trappist-common/src/impls.rs

```diff
@@ -0,0 +1,326 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Auxiliary struct/enums for parachain runtimes.
+//! Taken from polkadot/runtime/common (at a21cd64) and adapted for parachains.
+
+use frame_support::traits::{
+	fungibles::{self, Balanced, CreditOf},
+	Contains, Currency, Get, Imbalance, OnUnbalanced,
+};
+use pallet_asset_tx_payment::HandleCredit;
+use sp_runtime::{traits::Zero, Permill};
+use sp_std::marker::PhantomData;
+use xcm::latest::{AssetId, Fungibility::Fungible, MultiAsset, MultiLocation};
+use xcm_executor::traits::FilterAssetLocation;
+
+/// Type alias to conveniently refer to the `Currency::NegativeImbalance` associated type.
+pub type NegativeImbalance<T> = <pallet_balances::Pallet<T> as Currency<
+	<T as frame_system::Config>::AccountId,
+>>::NegativeImbalance;
+
+/// Type alias to conveniently refer to `frame_system`'s `Config::AccountId`.
+pub type AccountIdOf<R> = <R as frame_system::Config>::AccountId;
+
+/// Logic for the author to get a portion of fees.
+pub struct ToAuthor<R>(PhantomData<R>);
+impl<R> OnUnbalanced<NegativeImbalance<R>> for ToAuthor<R>
+where
+	R: pallet_balances::Config + pallet_collator_selection::Config,
+	AccountIdOf<R>:
+		From<polkadot_primitives::v2::AccountId> + Into<polkadot_primitives::v2::AccountId>,
+	<R as frame_system::Config>::RuntimeEvent: From<pallet_balances::Event<R>>,
+{
+	fn on_nonzero_unbalanced(amount: NegativeImbalance<R>) {
+		let author = <pallet_collator_selection::Pallet<R>>::account_id();
+		<pallet_balances::Pallet<R>>::resolve_creating(&author, amount);
+	}
+}
+
+pub struct DealWithFees<R>(PhantomData<R>);
+impl<R> OnUnbalanced<NegativeImbalance<R>> for DealWithFees<R>
+where
+	R: pallet_balances::Config + pallet_collator_selection::Config + pallet_treasury::Config,
+	pallet_treasury::Pallet<R>: OnUnbalanced<NegativeImbalance<R>>,
+	AccountIdOf<R>:
+		From<polkadot_primitives::v2::AccountId> + Into<polkadot_primitives::v2::AccountId>,
+	<R as frame_system::Config>::RuntimeEvent: From<pallet_balances::Event<R>>,
+{
+	fn on_unbalanceds<B>(mut fees_then_tips: impl Iterator<Item = NegativeImbalance<R>>) {
+		if let Some(fees) = fees_then_tips.next() {
+			// for fees, 80% to treasury, 20% to author
+			let mut split = fees.ration(80, 20);
+			if let Some(tips) = fees_then_tips.next() {
+				// for tips, if any, 100% to author
+				tips.merge_into(&mut split.1);
+			}
+			use pallet_treasury::Pallet as Treasury;
+			<Treasury<R> as OnUnbalanced<_>>::on_unbalanced(split.0);
+			<ToAuthor<R> as OnUnbalanced<_>>::on_unbalanced(split.1);
+		}
+	}
+}
+
+/// A `HandleCredit` implementation that naively transfers the fees to the block author.
+/// Will drop and burn the assets in case the transfer fails.
+pub struct AssetsToBlockAuthor<R>(PhantomData<R>);
```

Comment:
These were taken from Statemine. At the moment they are not used but they look interesting and we can consider in future iterations to use them.

Reply:
I wondered why the code was copied ? ... instead of referencing existing crates (e.g. parachain-commons)

Reply:
I like having them in our code so we keep the flexibility to perform any custom logic and do not depend on Statemine at the time to support this impls. What do you think? do you think that is more beneficial to referencing the existing crates?

Reply:
Yes, if there's no customization yet, I find it overkill to have a `trappist-common` crate. We might as well reference `parachains-common`, that makes the code lineage very clear (i.e. where is this logic (= trait implementations) coming from). If `AssetsToBlockAuthor` (which is coming from Statemine) looks interesting but is unused for now, just take a note of it, but don't include it in the code.

Reply:
I will remove the traits `AssetsToBlockAuthor`, `NonZeroIssuance` and `AssetsFrom` as they are not used now. However i realized that we still have to keep the custom implementation of `DealWithFees` in the `trappist-common` crate since the `DealWithFees` from `parachains-common` does not split the 80/20 of the fees as we agreed.

Reply:
Put it in the Trappist runtime's for now (in an `impls.rs' file). Let's keep things simple.

Reply:
Addressed.